drivers/espi/espi_handlers.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/espi.h>
+#include <syscall_handler.h>
+
+
+static inline int z_vrfy_espi_config(struct device *dev, struct espi_cfg *cfg)
+{
+	struct  espi_cfg cfg_copy;
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, config));
+	Z_OOPS(z_user_from_copy(&cfg_copy, cfg,
+				sizeof(struct espi_cfg)));
+
+	return z_impl_espi_config(dev, &cfg_copy);
+}
+#include <syscalls/espi_config_mrsh.c>
+
+static inline bool z_vrfy_espi_get_channel_status(struct device *dev,
+						  enum espi_channel ch)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, get_channel_status));
+
+	return z_impl_espi_get_channel_status(dev, ch);
+}
+#include <syscalls/espi_get_channel_status_mrsh.c>
+
+static inline int z_vrfy_espi_read_lpc_request(struct device *dev,
+					       enum lpc_peripheral_opcode op,
+					       u32_t *data)
+{
+	int ret;
+	u32_t data_copy;
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, read_lpc_request));
+	Z_OOPS(Z_SYSCALL_MEMORY_WRITE(data, sizeof(*data)));
+
+	ret = z_impl_espi_read_lpc_request(dev, op, &data_copy);
+
+	*data = data_copy;
```

Comment:
should just do z_user_to_copy() here and omit the Z_SYSCALL_MEMORY_WRITE above.

Reply:
@andrewboie  if I do that I'll need to call `z_impl_espi_read_lpc_request` even if the given memory was invalid, because the check will occur after the call. It seems better to check the memory before the call so we don't execute it if the parameter is invalid. Don't you think ?

Reply:
Is that actually a requirement? Please think about this, I don't actually believe it is.

Consider later modifications to support SMAP where user memory access is strictly controlled through accessor functions instead of `*data = data_copy`

Reply:
that is not a requirement *I think*, though it is a possible waste of resource. Though I'm fine we move it to use the copy function. But definitely will be cases where a function should not be executed if a given memory is invalid. In that case we will require the memory access check + `z_user_to_copy`.

Reply:
> But definitely will be cases where a function should not be executed if a given memory is invalid. In that case we will require the memory access check + z_user_to_copy.

Can you provide an example of this?

Reply:
I can come up with other question, can you ensure that this will never be necessary ? I can imagine a driver (or a generic code) that change status everytime is called. Entropy for instance, you can probably starve an internal buffer calling it with a wrong memory that will never be used.

Is that a real case / threat, probably not, but it seems suspicious execute a call that shouldn't be executed because it was given wrong parameters.
I can understand SMAP argument and since we are not discussing concrete cases I'll change it, but we should keep an eye on this possible case IMHO.

Reply:
To me it is not problematic for an API call to be executed, succeed, and then a memory error generated when attempting to copy some result back to memory accessible by user mode. I don't see it fundamentally different from having a coding mistake the next line after the API call which then induces an error. These kinds of issues are always coding mistakes which is why we Z_OOPS().

Reply:
pr changed, can you look it again ?